### ISSUE TITLE
snis_client segfaulted after warp gate jump

### ISSUE THREAD

**Issue by smcameron** [smcameron/space-nerds-in-space#160]
Saw this this morning.

From /var/log/kern.log

```
Sep 16 09:49:48 wombat kernel: [ 6809.248896] snis_client[14897]: segfault at ab000925ddc ip 0000000000488f71 sp 00007ffe800e19d0 error 4 in snis_client[400000+108000]
```

And then noticed:

```
wombat log # grep snis kern.log.1
Sep 10 12:59:25 wombat kernel: [15396.599553] snis_client[13879]: segfault at 8 ip 000000000041f804 sp 00007fff9a23d7a8 error 4 in snis_client[400000+ee000]
Sep 10 13:00:53 wombat kernel: [15484.700854] snis_client[14552]: segfault at 8 ip 00000000004281e3 sp 00007fffd67d9c00 error 4 in snis_client[400000+109000]
Sep 15 11:22:10 wombat kernel: [11951.991397] snis_multiverse[15277]: segfault at 8ed038 ip 00007fd60fd1858a sp 00007ffd5d602818 error 4 in libc-2.23.so[7fd60fc79000+1c0000]
Sep 15 11:22:53 wombat kernel: [11995.054018] snis_multiverse[15373]: segfault at 8ed038 ip 00007fa55b46e58a sp 00007ffcd3a9aea8 error 4 in libc-2.23.so[7fa55b3cf000+1c0000]
wombat log #
```

The snis_multiverse ones on Sep 15th were likely while I was working on autowrangler stuff, so I'm not too concerned about those unless I start seeing them again.

The snis_client ones on Sep 10th, I have no idea about.

The snis_client one on Sep 16th happened shortly after a warp gate jump (many previous successful warp gate jumps in the same process)... really don't have much in the way of a clue though.

**Comment by smcameron** [comment]
A possible clue:

```
(gdb) info symbol 0x0000000000488f71
show_death_screen + 800 in section .text
```

I think this:

```
0x488f5c <show_death_screen+779>        mov    0x284516(%rip),%eax        # 0x70d478 <my_ship_oid>
0x488f62 <show_death_screen+785>        mov    %eax,%eax
0x488f64 <show_death_screen+787>        imul   $0xab0,%rax,%rax
0x488f6b <show_death_screen+794>        add    $0x92688c,%rax
0x488f71 <show_death_screen+800>        mov    (%rax),%edx
0x488f73 <show_death_screen+802>        lea    -0x70(%rbp),%rax
0x488f77 <show_death_screen+806>        mov    $0x4ee175,%esi
0x488f7c <show_death_screen+811>        mov    %rax,%rdi
0x488f7f <show_death_screen+814>        mov    $0x0,%eax
0x488f84 <show_death_screen+819>        callq  0x4100a0 <sprintf@plt>
0x488f89 <show_death_screen+824>        mov    $0x1f4,%edi
0x488f8e <show_death_screen+829>        callq  0x466690 <txy>
0x488f93 <show_death_screen+834>        pxor   %xmm3,%xmm3

```

Corresponds to this, so it is setting up to do the call to sprintf on 10634, and blowing up.

```
10634                   sprintf(buf, "RESPAWNING IN %d SECONDS", go[my_ship_oid].respawn_time); 10635                   sng_abs_xy_draw_string(buf, TINY_FONT, txx(20), txy(500));  
10636           }
```

So I expect maybe my_ship_oid is not correct somehow?

**Comment by smcameron** [comment]
Ok I think I know what happened. the poorly named my_ship_oid, which is really an index into go[], not an object id as the name would imply, can get reset to -1 here:

```
static int process_client_id_packet(void)
{
        unsigned char buffer[100];
        uint32_t id;
        int rc;

        assert(sizeof(buffer) > sizeof(struct client_ship_id_packet) - sizeof(uint8_t));
        rc = read_and_unpack_buffer(buffer, "w", &id);
        if (rc)
                return rc;
        my_ship_id = id;
        my_ship_oid = UNKNOWN_ID; <------------------- HERE
```

That gets called whenever a client connects to a snis_server process, the snis_server process sends the client its object ID, however, the player ship object itself will not yet have been added to go[]. On a fresh start of snis_client, I think this doesn't come up as a problem, but after a warp gate jump, it can come up as a problem.

So, that means we need to be extra careful whenever we try to access the player object, lest we access it between the time that process_client_id_packet() is called and the time that the player object is added into the  go[] array.

**Comment by smcameron** [comment]
This particular crash should be fixed by 75d0f4dcd9f7a592a749eb12a4f4be9be54267f6
